File: src/RoboTasker.Api/Models/Tasks/CreateTaskRequest.cs
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using RoboTasker.Application.Robots.Tasks.CreateTask;
using RoboTasker.Domain.Tasks.Data;

namespace RoboTasker.Api.Models.Tasks;

public class CreateTaskRequest
{
    public string Name { get; set; } = null!;
    public string? Description { get; set; }
    
    public TimeSpan EstimatedDuration { get; set; }
    public int Priority { get; set; }
    public double Complexity { get; set; }
    public Guid CategoryId { get; set; }

    [FromForm(Name = "requirements")]
    public string? RequirementsJson { get; set; } 
    
    [FromForm(Name = "data")]
    public string? DataJson { get; set; } 
    
    [JsonIgnore]
    public IList<CreateTaskRequirementCommand>? Requirements => JsonConvert.DeserializeObject<IList<CreateTaskRequirementCommand>>(RequirementsJson?? "[]");

    [JsonIgnore]
    public IList<CreateTaskDataRequest>? Data => JsonConvert.DeserializeObject<IList<CreateTaskDataRequest>>(DataJson?? "[]");
    
    [FromForm]
    public IFormFileCollection? Files { get; set; }
}

public class CreateTaskDataRequest
{
    public string Key { get; set; } = null!;
    public RobotTaskDataType Type { get; set; }
    
    public object? Value { get; set; } 
    public Guid? ExistingId { get; set; }
}

File: src/Nerbotix.UI/src/app/models/robots/tasks/requests/create-task-request.ts
import {TaskRequirementLevel} from '../task-requirement-level';
import {TaskDataType} from '../task-data-type';
import {Guid} from 'guid-typescript';

export interface CreateTaskRequest {
  name: string;
  description?: string;
  estimatedDuration: string;
  priority: number;
  categoryId?: Guid;
  complexity: number;
  requirements?: CreateTaskRequirementRequest[];
  data?: CreateTaskDataRequest[];
  files?: File[];
}

export interface CreateTaskRequirementRequest {
  capabilityId: string;
  level: TaskRequirementLevel;
  existingId?: Guid;
}

export interface CreateTaskDataRequest {
  key: string;
  type: TaskDataType;
  value?: any;
  existingId?: Guid;
}
